// @logicflow/core/lib/model/EditConfigModel.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EditConfigModel = void 0;
var lodash_es_1 = require("lodash-es");
var mobx_1 = require("../util/mobx");
var SilentConfig = {
    stopZoomGraph: false,
    stopScrollGraph: false,
    stopMoveGraph: false,
    adjustEdge: false,
    adjustEdgeStartAndEnd: false,
    adjustNodePosition: false,
    hideAnchors: true,
    allowRotate: false,
    allowResize: false,
    nodeSelectedOutline: true,
    nodeTextEdit: false,
    edgeTextEdit: false,
    nodeTextDraggable: false,
    edgeTextDraggable: false,
};
var keys = [
    'isSilentMode',
    'stopZoomGraph',
    'stopScrollGraph',
    'stopMoveGraph',
    'adjustEdge',
    'adjustEdgeMiddle',
    'adjustEdgeStartAndEnd',
    'adjustNodePosition',
    'hideAnchors',
    'allowRotate',
    'allowResize',
    'hoverOutline',
    'nodeSelectedOutline',
    'edgeSelectedOutline',
    'nodeTextEdit',
    'edgeTextEdit',
    'nodeTextDraggable',
    'edgeTextDraggable',
    'multipleSelectKey',
    'autoExpand',
];
/**
 * 页面编辑配置
 */
var EditConfigModel = /** @class */ (function () {
    function EditConfigModel(config) {
        this.isSilentMode = false;
        this.stopZoomGraph = false;
        this.stopScrollGraph = false;
        this.stopMoveGraph = false;
        this.adjustEdge = true;
        this.adjustEdgeMiddle = false;
        this.adjustEdgeStartAndEnd = false;
        this.adjustNodePosition = true;
        this.hideAnchors = false;
        this.allowRotate = false;
        this.allowResize = false;
        this.hoverOutline = true;
        this.nodeSelectedOutline = true;
        this.edgeSelectedOutline = true;
        this.nodeTextEdit = true;
        this.edgeTextEdit = true;
        this.nodeTextDraggable = false;
        this.edgeTextDraggable = false;
        this.autoExpand = false;
        this.multipleSelectKey = '';
        this.defaultConfig = {}; // 设置为静默模式之前的配置，在取消静默模式后恢复
        (0, lodash_es_1.assign)(this, this.getConfigDetail(config));
    }
    EditConfigModel.prototype.updateEditConfig = function (config) {
        var newConfig = this.getConfigDetail(config);
        (0, lodash_es_1.assign)(this, newConfig);
    };
    EditConfigModel.prototype.getConfigDetail = function (config) {
        var isSilentMode = config.isSilentMode, textEdit = config.textEdit;
        var conf = {};
        // false表示从静默模式恢复
        if (isSilentMode === false) {
            (0, lodash_es_1.assign)(conf, this.defaultConfig);
        }
        // 如果不传，默认undefined表示非静默模式
        if (isSilentMode === true && isSilentMode !== this.isSilentMode) {
            // https://github.com/didi/LogicFlow/issues/1180
            // 如果重复调用isSilentMode=true多次，会导致this.defaultConfig状态保存错误：保存为修改之后的Config
            // 因此需要阻止重复赋值为true，使用config.isSilentMode !== this.isSilentMode
            var silentConfig = (0, lodash_es_1.pick)(SilentConfig, keys);
            // 在修改之前，
            this.defaultConfig = {
                stopZoomGraph: this.stopZoomGraph,
                stopScrollGraph: this.stopScrollGraph,
                stopMoveGraph: this.stopMoveGraph,
                adjustEdge: this.adjustEdge,
                adjustEdgeMiddle: this.adjustEdgeMiddle,
                adjustEdgeStartAndEnd: this.adjustEdgeStartAndEnd,
                adjustNodePosition: this.adjustNodePosition,
                hideAnchors: this.hideAnchors,
                allowRotate: this.allowRotate,
                allowResize: this.allowResize,
                hoverOutline: this.hoverOutline,
                nodeSelectedOutline: this.nodeSelectedOutline,
                edgeSelectedOutline: this.edgeSelectedOutline,
                nodeTextEdit: this.nodeTextEdit,
                edgeTextEdit: this.edgeTextEdit,
                nodeTextDraggable: this.nodeTextDraggable,
                edgeTextDraggable: this.edgeTextDraggable,
                autoExpand: this.autoExpand,
            };
            (0, lodash_es_1.assign)(conf, silentConfig);
        }
        // 如果不传，默认undefined表示允许文本编辑
        if (textEdit === false) {
            (0, lodash_es_1.assign)(conf, {
                nodeTextEdit: false,
                edgeTextEdit: false,
            });
        }
        var userConfig = (0, lodash_es_1.pick)(config, keys);
        return (0, lodash_es_1.assign)(conf, userConfig);
    };
    EditConfigModel.prototype.getConfig = function () {
        return (0, lodash_es_1.pick)(this, keys);
    };
    __decorate([
        mobx_1.observable
    ], EditConfigModel.prototype, "isSilentMode", void 0);
    __decorate([
        mobx_1.observable
    ], EditConfigModel.prototype, "stopZoomGraph", void 0);
    __decorate([
        mobx_1.observable
    ], EditConfigModel.prototype, "stopScrollGraph", void 0);
    __decorate([
        mobx_1.observable
    ], EditConfigModel.prototype, "stopMoveGraph", void 0);
    __decorate([
        mobx_1.observable
    ], EditConfigModel.prototype, "adjustEdge", void 0);
    __decorate([
        mobx_1.observable
    ], EditConfigModel.prototype, "adjustEdgeMiddle", void 0);
    __decorate([
        mobx_1.observable
    ], EditConfigModel.prototype, "adjustEdgeStartAndEnd", void 0);
    __decorate([
        mobx_1.observable
    ], EditConfigModel.prototype, "adjustNodePosition", void 0);
    __decorate([
        mobx_1.observable
    ], EditConfigModel.prototype, "hideAnchors", void 0);
    __decorate([
        mobx_1.observable
    ], EditConfigModel.prototype, "allowRotate", void 0);
    __decorate([
        mobx_1.observable
    ], EditConfigModel.prototype, "allowResize", void 0);
    __decorate([
        mobx_1.observable
    ], EditConfigModel.prototype, "hoverOutline", void 0);
    __decorate([
        mobx_1.observable
    ], EditConfigModel.prototype, "nodeSelectedOutline", void 0);
    __decorate([
        mobx_1.observable
    ], EditConfigModel.prototype, "edgeSelectedOutline", void 0);
    __decorate([
        mobx_1.observable
    ], EditConfigModel.prototype, "nodeTextEdit", void 0);
    __decorate([
        mobx_1.observable
    ], EditConfigModel.prototype, "edgeTextEdit", void 0);
    __decorate([
        mobx_1.observable
    ], EditConfigModel.prototype, "nodeTextDraggable", void 0);
    __decorate([
        mobx_1.observable
    ], EditConfigModel.prototype, "edgeTextDraggable", void 0);
    __decorate([
        mobx_1.observable
    ], EditConfigModel.prototype, "autoExpand", void 0);
    __decorate([
        mobx_1.action
    ], EditConfigModel.prototype, "updateEditConfig", null);
    return EditConfigModel;
}());
exports.EditConfigModel = EditConfigModel;
exports.default = EditConfigModel;
